subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -796,20 +789,35 @@ func generatePodScheduleResult(
 			PodPreemptInfo: &internal.PodPreemptInfo{VictimPods: victimPods},
 		}
 	} else {
-		// we check suggested nodes after the preemption is done, otherwise the preemption victims
+		// we find the selected node after the preemption is done, otherwise the preemption victims
 		// may cause the selected node to be excluded from the suggested nodes
-		// TODO: Keep selectedNode within suggestedNodeSet, so here should be Unreachable
-		if !suggestedNodeSet.Contains(selectedNode) && newGroup {
-			panic(fmt.Sprintf("[%v]: node %v picked by algorithm but not in K8S candidates",
-				internal.Key(pod), selectedNode))
+		affinityGroupBindInfo, selectedNode, selectedGpuIndices := generateAffinityGroupBindInfo(
+			groupPhysicalPlacement, groupVirtualPlacement, cellLevelToType, currentGpuNum, currentPodIndex, suggestedNodeSet)
+		if affinityGroupBindInfo == nil {
+			reason := "insufficient capacity in physical cluster"
+			if groupVirtualPlacement != nil {
+				reason = fmt.Sprintf("insufficient quota in VC %v", vc)
+			}
+			return internal.PodScheduleResult{
+				PodWaitInfo: &internal.PodWaitInfo{
+					Reason: reason,
+				},
+			}
+		}
+		if selectedNode == "" {
+			return internal.PodScheduleResult{
```

Comment:
Still consider newGroup, so
  if selectedNode == "" && newGroup {

Reply:
if newGroup == false, selectedNode must be "". See collectPreemptionVictims

---
In reply to: [353695916](https://github.com/microsoft/pai/pull/3962#discussion_r353695916) [](ancestors = 353695916)

Reply:
So for line 876, it should be 

  if podGpuNum == currentGpuNum && (suggestedNodeSet.Contains(mbi.PodPlacements[currentPodIndex].PhysicalNode) || !newGroup) {

right?



---
In reply to: [354073449](https://github.com/microsoft/pai/pull/3962#discussion_r354073449) [](ancestors = 354073449,353695916)